src/libNode/Node.cpp

```diff
@@ -271,43 +228,254 @@ void Node::Prepare(bool runInitializeGenesisBlocks) {
       m_mediator.m_dsBlockChain.GetLastBlock().GetHeader().GetBlockNum() + 1);
 }
 
-bool Node::StartRetrieveHistory() {
+bool Node::StartRetrieveHistory(bool& wakeupForUpgrade) {
   LOG_MARKER();
 
   m_mediator.m_txBlockChain.Reset();
   m_mediator.m_dsBlockChain.Reset();
-
   m_retriever = make_shared<Retriever>(m_mediator);
 
-  bool ds_result;
-  std::thread tDS(&Retriever::RetrieveDSBlocks, m_retriever.get(),
-                  std::ref(ds_result));
-  // retriever->RetrieveDSBlocks(ds_result);
+  if (LOOKUP_NODE_MODE) {
+    m_mediator.m_DSCommittee->clear();
+  }
 
-  bool tx_result;
-  std::thread tTx(&Retriever::RetrieveTxBlocks, m_retriever.get(),
-                  std::ref(tx_result));
-  // retriever->RetrieveTxBlocks(tx_result);
+  BlockStorage::GetBlockStorage().GetDSCommittee(
+      m_mediator.m_DSCommittee, m_mediator.m_ds->m_consensusLeaderID);
+
+  std::vector<unsigned char> metaRes;
+  if (BlockStorage::GetBlockStorage().GetMetadata(MetaType::WAKEUPFORUPGRADE,
+                                                  metaRes)) {
+    if (metaRes[0] == '1') {
+      wakeupForUpgrade = true;
+      BlockStorage::GetBlockStorage().PutMetadata(MetaType::WAKEUPFORUPGRADE,
+                                                  {'0'});
+    }
+  }
 
+  /// Retrieve Tx blocks, relative final-block state-delta from persistence
   bool st_result = m_retriever->RetrieveStates();
+  bool tx_result;
+  m_retriever->RetrieveTxBlocks(tx_result, wakeupForUpgrade);
+
+  if (!tx_result) {
+    return false;
+  }
+
+  /// Retrieve lacked Tx blocks from lookup nodes
+  uint64_t oldTxNum = m_mediator.m_txBlockChain.GetBlockCount();
+  m_synchronizer.FetchLatestTxBlocks(m_mediator.m_lookup,
+                                     m_mediator.m_txBlockChain.GetBlockCount());
+  LOG_GENERAL(INFO, "Retrieve final block from lookup node, please wait...");
 
-  tDS.join();
-  tTx.join();
+  unique_lock<mutex> lock(m_mediator.m_lookup->m_MutexCVSetTxBlockFromSeed);
+  if (m_mediator.m_lookup->cv_setTxBlockFromSeed.wait_for(
+          lock, chrono::seconds(RECOVERY_SYNC_TIMEOUT)) == cv_status::timeout) {
+    LOG_GENERAL(WARNING,
+                "GetTxBlockFromLookupNodes Timeout, cannot get final block "
+                "from lookup node!");
```

Comment:
If failed should fetch again. also better to refer to the logic of `StartSynchronization` to get Offline Lookups first